kroxylicious-filters/kroxylicious-authorization/src/main/java/io/kroxylicious/filter/authorization/CreatePartitionsEnforcement.java

```diff
@@ -0,0 +1,82 @@
+ */
+
+package io.kroxylicious.filter.authorization;
+
+import java.util.concurrent.CompletionStage;
+
+import org.apache.kafka.common.message.CreatePartitionsRequestData;
+import org.apache.kafka.common.message.CreatePartitionsResponseData;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.message.ResponseHeaderData;
+import org.apache.kafka.common.protocol.Errors;
+
+import io.kroxylicious.authorizer.service.Action;
+import io.kroxylicious.authorizer.service.Decision;
+import io.kroxylicious.proxy.filter.FilterContext;
+import io.kroxylicious.proxy.filter.RequestFilterResult;
+
+class CreatePartitionsEnforcement extends ApiEnforcement<CreatePartitionsRequestData, CreatePartitionsResponseData> {
+
+    @Override
+    short minSupportedVersion() {
+        return 0;
+    }
+
+    @Override
+    short maxSupportedVersion() {
+        return 3;
+    }
+
+    CompletionStage<RequestFilterResult> onRequest(RequestHeaderData header,
+                                                   CreatePartitionsRequestData request,
+                                                   FilterContext context,
+                                                   AuthorizationFilter authorizationFilter) {
+        var actions = request.topics().stream()
+                .map(cpd -> new Action(TopicResource.ALTER, cpd.name()))
+                .toList();
+        return authorizationFilter.authorization(context, actions)
+                .thenCompose(authorization -> {
+                    var decisions = authorization.partition(request.topics(),
+                            TopicResource.ALTER,
+                            CreatePartitionsRequestData.CreatePartitionsTopic::name);
+                    if (decisions.get(Decision.ALLOW).isEmpty()) {
+                        // Shortcircuit if there's no allowed topics
+                        var creatableTopics = decisions.get(Decision.DENY).stream()
+                                .map(CreatePartitionsEnforcement::topicAuthzFailed)
+                                .toList();
+                        return context.requestFilterResultBuilder().shortCircuitResponse(
+                                new ResponseHeaderData().setCorrelationId(header.correlationId()),
+                                new CreatePartitionsResponseData().setResults(creatableTopics)).completed();
+                    }
+                    else if (decisions.get(Decision.DENY).isEmpty()) {
+                        // Just forward if there's no denied topics
+                        return context.forwardRequest(header, request);
+                    }
+                    else {
+                        var topicCollection = new CreatePartitionsRequestData.CreatePartitionsTopicCollection();
+                        for (var topic : decisions.get(Decision.ALLOW)) {
+                            topicCollection.mustAdd(topic.duplicate());
```

Comment:
Curiousity - why do we duplicate?

Reply:
The kafka API sometimes uses linked collections where the elements contain the references. If you add an element from another collection, the add can fail due to the state of the object you are adding.

We use `mustAdd` for this reason, it throws if you try to add an element in a bad state.

It's a nasty bit of API, they look like harmless collections.

Reply:
I assume we're duplicating for similar reasons, to discard the link state

Reply:
> If you add an element from another collection, the add can fail due to the state of the object you are adding.

Oh yes.. I've been bitten by this in the User Namespace Filter work!

I see that duplicate drops the next/prevs.  Thank you.